frontend/lib/resolution/Resolver.cpp

```diff
@@ -4050,174 +4173,386 @@ void Resolver::exit(const New* node) {
-        QualifiedType eltType;
-        if (loop->numStmts() == 1) {
-          eltType = byPostorder.byAst(loop->stmt(0)).type();
-        } else if (loop->numStmts() == 0) {
-          eltType = QualifiedType(QualifiedType::TYPE, AnyType::get(context));
-        } else {
-          CHPL_ASSERT(false && "array expression with multiple loop body statements?");
-        }
+    // TODO: resolve array types when the iterand is something other than
+    // a domain.
+    if (eltType.isType() || eltType.kind() == QualifiedType::TYPE_QUERY) {
+      eltType = QualifiedType(QualifiedType::TYPE, eltType.type());
+      auto arrayType = ArrayType::getArrayType(context, iterandType, eltType);
+
+      auto& re = rv.byPostorder.byAst(loop);
+      re.setType(QualifiedType(QualifiedType::TYPE, arrayType));
+    }
+  }
+}
 
-        // TODO: resolve array types when the iterand is something other than
-        // a domain.
-        if (eltType.isType() || eltType.kind() == QualifiedType::TYPE_QUERY) {
-          eltType = QualifiedType(QualifiedType::TYPE, eltType.type());
-          auto arrayType = ArrayType::getArrayType(context, iterandType, eltType);
+bool Resolver::enter(const IndexableLoop* loop) {
+  auto forLoop = loop->toFor();
+  bool isParamForLoop = forLoop != nullptr && forLoop->isParam();
 
-          ResolvedExpression& re = byPostorder.byAst(loop);
-          re.setType(QualifiedType(QualifiedType::TYPE, arrayType));
-        }
-      }
+  // whether this is a param or regular loop, before entering its body
+  // or considering its iterand, resolve expressions in the loop's attribute
+  // group.
+  if (auto ag = loop->attributeGroup()) {
+    ag->traverse(*this);
+  }
+
+  if (isParamForLoop) return resolveParamForLoop(*this, loop->toFor());
+
+  auto iterand = loop->iterand();
+  QualifiedType idxType;
+
+  // The 'zip' visitor will consult the parent loop in order to resolve.
+  if (iterand->isZip()) {
+    iterand->traverse(*this);
+    idxType = byPostorder.byAst(iterand).type();
+
+  // Otherwise - configure a loop-specific policy and resolve the iterator.
+  } else {
+    int m = IterDetails::NONE;
+    if (!loop->isForall()) m |= IterDetails::SERIAL;
+    if (loop->isBracketLoop() || loop->isForall()) {
+      m |= IterDetails::STANDALONE | IterDetails::LEADER_FOLLOWER;
     }
+
+    CHPL_ASSERT(m != IterDetails::NONE);
+
+    auto dt = resolveIterDetails(*this, loop, iterand, {}, m);
+
+    idxType = dt.idxType;
   }
 
+  enterScope(loop);
+
+  if (const Decl* idx = loop->index()) {
+    ResolvedExpression& re = byPostorder.byAst(idx);
+    re.setType(idxType);
+  }
+
+  if (auto with = loop->withClause()) {
+    with->traverse(*this);
+  }
+
+  loop->body()->traverse(*this);
+
+  // TODO: What would it take to make backpatching of array types happen
+  // _before_ / without resolving an iterator? Currently we rely on iterator
+  // resolution to resolve the iterand for us.
```

Comment:
I thought that we only needed to detect if elements yielded by the loop expression are types rather than values? Also it checks something about there being a domain involved. Seems like it'd be easy enough to filter these earlier. Am I wrong about something?

Reply:
I'll change the stale `TODO`.